src/threadinterrupt.h

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2016 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <mutex>
+#include <condition_variable>
+#include <atomic>
+
+class CThreadInterrupt
+{
+public:
+    CThreadInterrupt(std::condition_variable& condIn, std::mutex& mutIn) : cond(condIn), mut(mutIn), val(false) {}
+    void reset()
+    {
+        val.store(false, std::memory_order_release);
+    }
+    void operator()()
+    {
+        {
+            std::unique_lock<std::mutex> lock(mut);
+            val.store(true, std::memory_order_release);
+        }
+        cond.notify_all();
+    }
+    explicit operator bool() const
+    {
+        return val.load(std::memory_order_acquire) == true;
+    }
+
+private:
+    std::condition_variable& cond;
+    std::mutex& mut;
+    std::atomic<bool> val;
+
+    template <typename Duration>
+    friend bool InterruptibleSleep(const Duration&, CThreadInterrupt&);
+};
+
+/*
+ * Sleep for the stated period of time, interruptible by clearing the flag and notifying the condvar.
+ * @param   rel_time maximum time to wait. Should be a std::chrono::duration.
+ * @param   threadInterrupt The interrupt that may wake the sleep
+ * @returns false if the sleep was interrupted, true otherwise
+ */
+template <typename Duration>
+bool InterruptibleSleep(const Duration& rel_time, CThreadInterrupt& threadInterrupt)
```

Comment:
Can we just move this into a member function of the class?

Also, does it need to be a template? I think if you set Duration to something that the various durations can be converted into it should do so intelligently. Would mean less stuff to compile in headers.